File: Java_Programming/Doubly_Linear_LL.java

class node
{
    public int data;
    public node next;
    public node prev;

    public node(int no)
    {
        this.data=no;
        this.next=null;
        this.prev=null;
    }

}

class DoublyLL
{
    private node first;
    private int iCount;

    public DoublyLL()
    {
        System.out.println("Object of Doubly CL is created");
        this.first=null;
        this.iCount=0;
    }

    public void InsertFirst(int no)
    {
        node newn=null;
        newn=new node(no);

        if(this.first==null)
        {
            this.first=newn;
        }
        else
        {
            newn.next=this.first;
            this.first.prev=newn;
            this.first=newn;
        }
        this.iCount++;
    }

    public void InsertLast(int no)
    {
        node temp=null;
        node newn=null;
        newn=new node(no);

        if(this.first==null)
        {
            this.first=newn;
        }
        else
        {
            temp=this.first;

            while(temp.next!=null)
            {
                temp=temp.next;
            }
            temp.next=newn;
            newn.prev=temp;
        }
        this.iCount++;
    }

    public void InsertAtPos(int no,int pos)
    {
        node newn=null;
        newn=new node(no);
        node temp=null;
        int iCnt=0;

        if((pos<1)||(pos>iCount+1))
        {
            System.out.println("Invalid position");
            return;
        }
        if(pos==1)
        {
            this.InsertFirst(no);
        }
        else if(pos==iCount+1)
        {
            this.InsertLast(no);
        }
        else
        {
            temp=this.first;

            for(iCnt=1;iCnt<pos-1;iCnt++)
            {
                temp=temp.next;
            }
            newn.next=temp.next;
            temp.next.prev=newn;
            temp.next=newn;

            this.iCount++;
        }
    }

    public void DeleteFirst()
    {
        if(this.first==null)
        {
            System.out.println("Linked list is empty");
            return;
        }
        else if(this.first.next==null)
        {
            this.first=null;
        }
        else
        {
            this.first=this.first.next;

        }

        System.gc();
        this.iCount--;
    }

    public void DeleteLast()
    {
        node temp=null;

        if(this.first==null)
        {
            System.out.println("Linked list is empty");
            return;
        }
        else if(this.first.next==null)
        {
            this.first=null;
        }
        else
        {
            temp=this.first;

            while(temp.next.next!=null)
            {
                temp=temp.next;
            }
            temp.next=null;
            
        }

        System.gc();
        this.iCount--;
    }

    public void DeleteAtPos(int pos)
    {
        node temp=null;
        int iCnt=0;

        if((pos<1)||(pos>iCount))
        {
            System.out.println("Invalid position");
            return;
        }
        if(pos==1)
        {
            this.DeleteFirst();
        }
        else if(pos==iCount)
        {
            this.DeleteLast();
        }
        else
        {
            temp=this.first;

            for(iCnt=1;iCnt<pos-1;iCnt++)
            {
                temp=temp.next;
            }
            temp.next=temp.next.next;

            System.gc();
            this.iCount--;
        }  
    }

    public void Display()
    {
        node temp=null;
        

        temp=this.first;

        while(temp!=null)
        {
            System.out.print("| "+temp.data+"|<=>");
            temp=temp.next;
        }
        System.out.println("null");

    }

    public int Count()
    {
        return this.iCount;
    }

}

class Doubly_Linear_LL
{
    public static void main(String Arr[])
    {
        DoublyLL obj=new DoublyLL();
        int iRet=0;

        obj.InsertFirst(84);
        obj.InsertFirst(27);
        obj.InsertFirst(93);

        obj.Display();
        iRet=obj.Count();
        System.out.println("Number of nodes in linkedlist are: "+iRet);
        System.out.println();

        obj.InsertLast(31);
        obj.InsertLast(25);
        obj.InsertLast(35);

        obj.Display();
        iRet=obj.Count();
        System.out.println("Number of nodes in linkedlist are: "+iRet);
        System.out.println();

        obj.DeleteFirst();

        obj.Display();
        iRet=obj.Count();
        System.out.println("Number of nodes in linkedlist are: "+iRet);
        System.out.println();

        obj.DeleteLast();

        obj.Display();
        iRet=obj.Count();
        System.out.println("Number of nodes in linkedlist are: "+iRet);
        System.out.println();

        obj.InsertAtPos(43,2);

        obj.Display();
        iRet=obj.Count();
        System.out.println("Number of nodes in linkedlist are: "+iRet);
        System.out.println();

        obj.DeleteAtPos(3);

        obj.Display();
        iRet=obj.Count();
        System.out.println("Number of nodes in linkedlist are: "+iRet);
        System.out.println();
    }
}

File: C++_Programming/Doubly_linear_LL.cpp
#include<iostream>

using namespace std;

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node * PNODE;

class DoublyLL
{
    private:
            PNODE first;
            int iCount;

    public:
            DoublyLL()
            {
                cout<<"Object of Doubly Linear LL is created\n";
                this->first=NULL;
                this->iCount=0;
            }

            void InsertFirst(int no)
            {
                PNODE newn=new NODE;
                
                newn->data=no;
                newn->next=NULL;
                newn->prev=NULL;

                if(this->first==NULL)
                {
                    this->first=newn;
                }
                else
                {
                    newn->next=this->first;
                    this->first=newn;
                }
                this->iCount++;
            }

            void InsertLast(int no)
            {
                PNODE newn=new NODE;
                PNODE temp=NULL;
                
                newn->data=no;
                newn->next=NULL;
                newn->prev=NULL;

                if(this->first==NULL)
                {
                    this->first=newn;
                }
                else
                {
                    temp=first;

                    while(temp->next!=NULL)
                    {
                        temp=temp->next;
                    }
                    temp->next=newn;
                    newn->prev=temp;
                }
                this->iCount++;
            }

            void DeleteFirst()
            {

                if(this->first==NULL)
                {
                    cout<<"Linked list is empty\n";
                    return;
                }
                else if(this->first->next==NULL)
                {
                    delete this->first;
                    this->first=NULL;
                }
                else
                {
                    this->first=this->first->next;
                    delete this->first->prev;
                    this->first->prev=NULL;
                }
                this->iCount--;
            }
            void DeleteLast()
            {
                PNODE temp=NULL;


                if(this->first==NULL)
                {
                    cout<<"Linked list is empty\n";
                    return;
                }
                else if(this->first->next==NULL)
                {
                    delete this->first;
                    this->first=NULL;
                }
                else
                {
                    temp=first;

                    while(temp->next->next!=NULL)
                    {
                        temp=temp->next;
                    }
                    
                    delete temp->next;
                    temp->next=NULL;

                }
                this->iCount--;
            }
            void Display()
            {
                PNODE temp=NULL;
                temp=first;

                while(temp!=NULL)
                {
                    cout<<"| "<<temp->data<<" |<=>";
                    temp=temp->next;
                }
                cout<<"NULL\n";
            }

            int Count()
            {

                return iCount;
            }

            void InsertAtPos(int no,int pos)
            {
                PNODE newn=new NODE;
                PNODE temp=NULL;
                int iCnt=0;

                if((pos<1)||(pos>this->iCount+1))
                {
                    cout<<"Inavlid position";
                    return;
                }
                if(pos==1)
                {
                    this->InsertFirst(no);
                }
                else if(pos==iCount+1)
                {
                    this->InsertLast(no);
                }
                else
                {
                    newn->data=no;
                    newn->next=NULL;
                    newn->prev=NULL;

                    temp=this->first;

                    for(iCnt=1;iCnt<pos-1;iCnt++)
                    {
                        temp=temp->next;
                    }
                    newn->next=temp->next;
                    newn->next->prev=newn;
                    temp->next=newn;

                    this->iCount++;
                }

            }

            void DeleteAtPos(int pos)
            {
                int iCnt=0;
                PNODE temp=NULL;
                PNODE target=NULL;

                if((pos<1)||(pos>this->iCount))
                {
                    cout<<"Inavlid position";
                    return;
                }
                if(pos==1)
                {
                    this->DeleteFirst();
                }
                else if(pos==iCount)
                {
                    this->DeleteLast();
                }
                else
                {
                    temp=this->first;

                    for(iCnt=1;iCnt<pos-1;iCnt++)
                    {
                        temp=temp->next;
                    }
                    target=temp->next;
                    temp->next=target->next;
                    target->next->prev=temp;
                    delete target;
                    
                    this->iCount--;
                }
            }
};

int main()
{

    DoublyLL obj;

    int iRet=0;

    obj.InsertFirst(43);
    obj.InsertFirst(93);
    obj.InsertFirst(27);

    obj.Display();
    iRet=obj.Count();
    cout<<"Number of nodes are:"<<iRet<<"\n";

    obj.InsertLast(84);
    obj.InsertLast(31);
    obj.InsertLast(25);

    obj.Display();
    iRet=obj.Count();
    cout<<"Number of nodes are:"<<iRet<<"\n";

    obj.DeleteFirst();
    

    obj.Display();
    iRet=obj.Count();
    cout<<"Number of nodes are:"<<iRet<<"\n";

    obj.DeleteLast();
    
    obj.Display();
    iRet=obj.Count();
    cout<<"Number of nodes are:"<<iRet<<"\n";

    obj.InsertAtPos(10,5);
    
    obj.Display();
    iRet=obj.Count();
    cout<<"Number of nodes are:"<<iRet<<"\n";

    obj.DeleteAtPos(5);
    

    obj.Display();
    iRet=obj.Count();
    cout<<"Number of nodes are:"<<iRet<<"\n";




    return 0;
}